### PR TITLE
ClientAI/BasicStrategy

### PR DESCRIPTION
**Pull request by pariciadra**
### This is the pull request of the basic AI.

> In this PR u gonna see implemented function used by the basic strategy and unimplemented functions used by other strategies.
> Including Documentation of each function & useful data used by it.

**Brief**
Currently AI is able to : 

1. get object need atm in the map
2. move forward if the object isn't present on the current tile
3. apply 2 strategies : survive & farm
4. every actions of AI is blocking, soon able to queue some requests to server
5. target a tile to take the required component to evolve, but can't reach it atm, need to queue before (pathfinding soon able)

**Need to fix** @31Nathan 
- [x] fillMap index 0 and 1 isn't working, i have to start at [index + 2] to find the mapTile[0] i don't know why
- [x] getTile(x) only and not getTile(x, y)
- [x] AttributeError: 'ClientCore' object has no attribute '_ClientCore__placesLeft'
- [x] Please add methods `def DecreaseFood(self, nbFood: int)` in Inventory class, i will need it to decrease the inventory food to know when the Ai need to go on survival mode
- [x] Same for every components for elevation

**Message for reviewers :** 

> U are welcome to indicate best way to do something, i'm just a beginner
> If u think u got a better idea of doing something, or simply if something wrong, please argument, explain, and when it's possible, propose an alternative.

**Just note that :** 

> The documentation is only filled in if it is deemed necessary, otherwise it is implicit and therefore useless. 
> Good reading

### CODE REVIEW

**Review by Cavonstavant — APPROVED** [review]

**Comment by Cavonstavant** [comment]
When defining a function in python, one can use the keyword `pass` in order to leave the body of the function to be develloped later

File: src/zappy_ai_src/ai_basic_strategy.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+    """----------------------------------------Getter and Setter for Core Strategy class-----------------------------"""
+
+    def __setAvailableSlots(self, availableSlots: int):
+        self.__availableSlots = availableSlots
+
+    def __setIsRunning(self, isRunning: bool):
+        self.__isRunning = isRunning
+
+    """
+        [ASKING], should i delete the old inventory and then re use the constructor the refill the inventory,
```
Comment (LucasTesnier) [inline src/zappy_ai_src/ai_basic_strategy.py:167]: I think the proper way is edit the current inventory object

**Review by LucasTesnier — APPROVED** [review]
NTR!

File: src/zappy_ai_src/ai_strategy_management.py [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+
+    def __setAvailableSlots(self, availableSlots: int):
+        self.__availableSlots = availableSlots
+
+    def __setIsRunning(self, isRunning: bool):
+        self.__isRunning = isRunning
+
+    """
+        [ASKING], should i delete the old inventory and then re use the constructor the refill the inventory,
+        or should we turn the method fill inventory as public
```
Comment (31Nathan) [inline src/zappy_ai_src/ai_strategy_management.py:210]: I think we should turn the object fill inventory as public, do it in this PR

File: src/zappy_ai_src/ai_strategy_management.py [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+        In case of delete, how can i delete object Inventory ?
+    """
+    def __setInventory(self, inventoryResponse: str):
+        self.__inventory = Inventory(inventoryResponse)
+
+    """
+        [ASKING], same as inventory
```
Comment (31Nathan) [inline src/zappy_ai_src/ai_strategy_management.py:217]: Same as Inventory, it's easier

**Review by 31Nathan — COMMENTED** [review]
Check my comments, will approve when everything is done

File: src/zappy_ai_src/ai_strategy_management.py (hunk 2 above)
Comment (pariciadra) [reply src/zappy_ai_src/ai_strategy_management.py:210]: yes good idea

File: src/zappy_ai_src/ai_strategy_management.py (hunk 3 above)
Comment (pariciadra) [reply src/zappy_ai_src/ai_strategy_management.py:217]: ok

**Review by LucasTesnier — APPROVED** [review]
NTR! GG

**Review by 31Nathan — APPROVED** [review]
LGTM !

File: src/zappy_ai_src/ai_function_wrapper.py [hunk 4]
```diff
@@ -86,9 +86,8 @@ def getNecessaryFunctions(self) -> bool:
             self.__GetRepPlaceObject = self.DLLibWrapper.getFunctionFromLibrary("c_interface_get_place_response")
             self.__GetRepIncantation = self.DLLibWrapper.getFunctionFromLibrary("c_interface_get_incantation_response")
             self.__GetNetworkState = self.DLLibWrapper.getFunctionFromLibrary("c_interface_get_network_state")
-            self.DLLibWrapper.closeLibray()
         except:
-            print("The provided lib doesn't contain all the required funtions.", file=stderr)
+            print("The provided lib doesn't contain all the required functions.", file=stderr)
```
Comment (Cavonstavant) [inline src/zappy_ai_src/ai_function_wrapper.py]: @31Nathan Altough this is descriptive, you really should consider giving more info
```suggestion
           except Execption e:
               printf(e) <-- Something like taht
               print("The provided lib doesn't contain all the required functions.", file=stderr)
```

File: src/zappy_ai_src/ai_strategy_management.py [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+        else:
+            self.__farming()
+        self.__actionsProceed()
+
+    def __actionsProceed(self):
+        """This is used to trigger actions depending on previous configuration of the strategy
+            Like getting the most required component at a time T
+        """
+        component = self.__getTargetComponent()
+        for i in range(0, self.__getPlayerMaxRange() + 1):
```
Comment (Cavonstavant) [inline src/zappy_ai_src/ai_strategy_management.py:312]: :+1:

**Review by Cavonstavant — APPROVED** [review]
This is great work man. You did get the hang of python, nice.
Documentation is clear and concise
THIS IS NICE!

**Pull request merged**